src/components/templates/Camp/components/MapRoomOverlay.tsx

```diff
@@ -0,0 +1,226 @@
+import React, { useCallback, useState } from "react";
+
+import { IS_BURN } from "secrets";
+import { retainAttendance } from "store/actions/Attendance";
+import {
+  getExtraLinkProps,
+  getRoomUrl,
+  isExternalUrl,
+  openRoomUrl,
+} from "utils/url";
+
+import { Attendances } from "types/Attendances";
+import { CampRoomData } from "types/CampRoomData";
+import { CampVenue } from "types/CampVenue";
+import { RoomVisibility } from "types/Venue";
+
+import { useDispatch } from "hooks/useDispatch";
+
+import CampAttendance from "./CampAttendance";
+
+interface MapRoomOverlayProps {
+  // Passed down from Map component
+  room: CampRoomData;
+  enterCampRoom: (room: CampRoomData) => void;
+
+  // Passed down from Camp component (via Map component)
+  venue: CampVenue;
+  attendances: Attendances;
+  selectedRoom?: CampRoomData;
+  setSelectedRoom: (room?: CampRoomData) => void;
+  setIsRoomModalOpen: (value: boolean) => void;
+}
+
+export const MapRoomOverlay: React.FC<MapRoomOverlayProps> = ({
+  enterCampRoom,
+  room,
+
+  venue,
+  attendances,
+  selectedRoom,
+  setSelectedRoom,
+  setIsRoomModalOpen,
+}) => {
+  const isSelectedRoom = room === selectedRoom;
+  const hasAttendance = attendances[`${venue.name}/${room.title}`];
+
+  const isRoomExternal = isExternalUrl(room.url);
+  const currentRoomUrl = getRoomUrl(room.url);
+
+  // TODO: Pass this down from Camp, maybe called selectRoom?
+  const selectRoom = useCallback(
+    (room: CampRoomData) => {
+      setSelectedRoom(room);
+      setIsRoomModalOpen(true);
+    },
+    [setIsRoomModalOpen, setSelectedRoom]
+  );
+
+  const [roomClicked, setRoomClicked] = useState<string | undefined>(undefined);
+
+  // TODO: can we just roll this logic into whatever is setting roomClicked, then rename above to isRoomClicked?
+  const isThisRoomClicked = roomClicked === room.title;
+
+  // TODO: I think this may be intended to toggle 'clicked' for a room
+  // TODO: what is a better semantic name for this?
+  const onClick1 = useCallback(() => {
+    if (!IS_BURN) {
+      selectRoom(room);
+    } else {
+      setRoomClicked((prevRoomClicked) =>
+        prevRoomClicked === room.title ? undefined : room.title
+      );
+    }
+  }, [room, selectRoom]);
+
+  const [isRoomHovered, setRoomHovered] = useState<CampRoomData | undefined>(
+    undefined
+  );
+
+  const dispatch = useDispatch();
+
+  const roomHovered = useCallback(() => {
+    dispatch(retainAttendance(true));
+    setRoomHovered(room);
+  }, [dispatch, room]);
+
+  const roomUnhovered = useCallback(() => {
+    dispatch(retainAttendance(false));
+    setRoomHovered(undefined);
+  }, [dispatch]);
+
+  // TODO: this is being spread here because the below code modifies it with .push()/etc
+  // const rooms = [...venue.rooms];
+
+  // TODO: what is the intent/purpose of this?
```

Comment:
The old code took a local copy of the array first, then modified it. So it could be trying to ensure the hovered/clicked/etc room is first in the array? Maybe?